sdk/eventhub/event-hubs/src/receiver.ts

```diff
@@ -301,49 +295,182 @@ export class EventHubConsumer {
+          await this._baseConsumer.close();
+          this._baseConsumer = void 0;
         }
       }
     } catch (err) {
-      log.error(
-        "[%s] An error occurred while closing the Receiver for %s: %O",
-        this._context.connectionId,
-        this._context.config.entityPath,
-        err
-      );
       throw err;
     } finally {
       this._isClosed = true;
     }
   }
 
-  private getCheckpoint(): number | undefined {
-    if (!this._streamingReceiver && !this._batchingReceiver) {
-      return;
-    }
-    let lastBatchingReceiverSequenceNum: number = -1;
-    let lastStreamingReceiverSequenceNum: number = -1;
-    if (this._batchingReceiver) {
-      lastBatchingReceiverSequenceNum = this._batchingReceiver.checkpoint;
-    }
-    if (this._streamingReceiver) {
-      lastStreamingReceiverSequenceNum = this._streamingReceiver.checkpoint;
+  private async _receiveBatch(
+    maxMessageCount: number,
+    maxWaitTimeInSeconds: number,
+    abortSignal?: AbortSignalLike
+  ): Promise<ReceivedEventData[]> {
+    // this shouldn't be able to happen since we
+    // already checked if the receiver was closed.
+    if (!this._baseConsumer) {
+      return [];
     }
 
-    const checkpoint = Math.max(lastStreamingReceiverSequenceNum, lastBatchingReceiverSequenceNum);
-    if (checkpoint === -1) {
-      return;
-    }
-    return checkpoint;
+    // store events across multiple retries
+    const receivedEvents: ReceivedEventData[] = [];
+
+    const retrieveEvents = (): Promise<ReceivedEventData[]> => {
+      return new Promise(async (resolve, reject) => {
+        let timer: any;
+        const logOnAbort = (): void => {
+          const baseConsumer = this._baseConsumer;
+          const name = baseConsumer && baseConsumer.name;
+          const address = baseConsumer && baseConsumer.address;
+          const desc: string =
+            `[${this._context.connectionId}] The request operation on the Receiver "${name}" with ` +
+            `address "${address}" has been cancelled by the user.`;
+          log.error(desc);
+        };
+
+        const onAbort = (): void => {
+          clearTimeout(timer);
+          logOnAbort();
+          if (this._baseConsumer) {
+            this._baseConsumer.abort();
+          } else {
+            return reject(new AbortError("The receive operation has been cancelled by the user."));
+          }
+        };
+
+        // operation has been cancelled, so exit immediately
+        if (abortSignal && abortSignal.aborted) {
+          logOnAbort();
+          return reject(new AbortError("The receive operation has been cancelled by the user."));
+        }
+
+        // if this consumer was closed, _baseConsumer might be undefined.
+        // resolve the operation's promise with the events collected thus far in case
+        // the promise hasn't already been resolved.
+        if (!this._baseConsumer) {
+          return resolve(receivedEvents);
+        }
```

Comment:
Yes, I can move this up. I originally was thinking that the `abortSignal` should take precedence, but I think it would be safer to return the results if we have them.